client.go

```diff
@@ -0,0 +1,198 @@
+	}
+	reqURL := baseURL
+	if rt.path != "" {
+		reqURL += rt.path
+	}
+	req, err := Marshal(reqURL, rt.method, params)
+	if err != nil {
+		return errgo.Mask(err)
+	}
+
+	// Actually make the request.
+	doer := c.Doer
+	if doer == nil {
+		doer = http.DefaultClient
+	}
+	var httpResp *http.Response
+	body := req.Body.(BytesReaderCloser)
+	if doer1, ok := doer.(DoerWithBody); ok && body.Len() > 0 {
+		req.Body = nil
+		httpResp, err = doer1.DoWithBody(req, body)
+	} else {
+		httpResp, err = doer.Do(req)
+	}
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+
+	// Return response directly if required.
+	if respPt, ok := resp.(**http.Response); ok {
+		*respPt = httpResp
+		return nil
+	}
+	defer httpResp.Body.Close()
+	if httpResp.StatusCode == http.StatusOK {
+		return unmarshalJSONResponse(httpResp, resp)
+	}
+
+	errUnmarshaler := c.UnmarshalError
+	if errUnmarshaler == nil {
+		errUnmarshaler = DefaultErrorUnmarshaler
+	}
+	err = errUnmarshaler(httpResp)
+	if err == nil {
+		err = errgo.Newf("unexpected HTTP response status: %s", httpResp.Status)
+	}
+	return err
+}
+
+// ErrorUnmarshaler returns a function which will unmarshal error
+// responses into new values of the same type as template. The argument
+// v must be a pointer. A new instance of it is created every time the
+// returned function is called.
+func ErrorUnmarshaler(template error) func(*http.Response) error {
+	t := reflect.TypeOf(template)
+	if t.Kind() != reflect.Ptr {
+		panic(errgo.Newf("cannot unmarshal errors into value of type %T", template))
+	}
+	t = t.Elem()
+	return func(resp *http.Response) error {
+		if 300 <= resp.StatusCode && resp.StatusCode < 400 {
+			// It's a redirection error.
+			loc, _ := resp.Location()
+			return fmt.Errorf("unexpected redirect (status %s) from %q to %q", resp.Status, resp.Request.URL, loc)
+		}
+		if err := checkIsJSON(resp.Header); err != nil {
+			return fmt.Errorf("cannot unmarshal error response (status %s): %v", resp.Status, err)
+		}
+		errv := reflect.New(t)
+		if err := unmarshalJSONResponse(resp, errv.Interface()); err != nil {
+			return fmt.Errorf("cannot unmarshal error response (status %s): %v", resp.Status, err)
+		}
+		return errv.Interface().(error)
+	}
+}
+
+func unmarshalJSONResponse(resp *http.Response, x interface{}) error {
+	// TODO export this function?
+
+	// Try to read all the body so that we can reuse the
+	// connection, but don't try *too* hard.
+	defer io.Copy(ioutil.Discard, io.LimitReader(resp.Body, 8*1024))
+	if x == nil {
+		return nil
+	}
+	if err := checkIsJSON(resp.Header); err != nil {
+		return errgo.Mask(err)
+	}
+	// Decode only a single JSON value, and then
+	// discard the rest of the body so that we can
+	// reuse the connection even if some foolish server
+	// has put garbage on the end.
+	dec := json.NewDecoder(resp.Body)
+	if err := dec.Decode(x); err != nil {
+		return errgo.Mask(err)
+	}
+	io.Copy(ioutil.Discard, resp.Body)
```

Comment:
Could you add a comment explaining why you want this and line 159, if indeed it is what you want? That will stop me in a few months thinking that's unnecessary and removing it.